### PR TITLE
Create a first prototype of auto-updates with Omaha 4 on macOS

### PR DESCRIPTION
**Pull request by mherrmann**
Resolves https://github.com/brave/brave-browser/issues/36946. The new implementation is hidden behind a default-disabled build flag, `brave_enable_updater`. So there should (TM) not be any user-visible changes yet.

The implementation is a prototype with the following prerequisites:

 1. The Omaha 4 update client must already be installed on the system.
 2. Brave must have been registered with O4.

Both are manual steps as of this writing. But when they are done, then Brave can be updated with Omaha 4 via `brave://settings/help`.

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

On macOS:

1. Please check that Brave's `.dmg` installer still works.
2. Please check that background updates still work.
3. Please check that component updates via `brave://components` still work.
4. Please check that extension updates via `brave://extensions` still work.

### CODE REVIEW

File: build/mac/BUILD.gn [hunk 1]
```diff
@@ -568,3 +585,17 @@ copy("copy_dmg_pkg") {
...
+if (brave_enable_updater) {
+  crx3("crx") {
+    inputs = [ target_dmg_path ]
+    output = "$root_out_dir/$chrome_product_full_name.crx3"
+    base_dir = get_path_info(target_dmg_path, "dir")
+
+    # The private key file is not in Git for obvious reasons. It needs to be
+    # copied into this directory manually. Changing the private key requires
+    # also changing kBravePublisherKeyHash in crx_verifier.cc.
+    key = "//brave/updater/crx-private-key.der"
```
Comment (goodov) [inline build/mac/BUILD.gn:598]: sec-review is required here. Not sure if it's right to put the private key during the build at this location.

File: patches/components-crx_file-BUILD.gn.patch [hunk 2]
```diff
@@ -0,0 +1,12 @@
+diff --git a/components/crx_file/BUILD.gn b/components/crx_file/BUILD.gn
+index 272b6f4f34e486804113475b9ae290b652236be8..b0fb4f603c196c86b1b6bffb275d03b70a789ded 100644
+--- a/components/crx_file/BUILD.gn
++++ b/components/crx_file/BUILD.gn
+@@ -75,7 +75,6 @@ proto_library("crx3_proto") {
+ 
+ if (host_toolchain == current_toolchain) {
+   executable("crx3_build_action") {
+-    testonly = true
```
Comment (goodov) [inline patches/components-crx_file-BUILD.gn.patch:9]: why this change was needed?

**Comment by iefremov** [comment]
perhaps @simonhong might be interested

File: build/mac/BUILD.gn (hunk 1 above)
Comment (mherrmann) [reply build/mac/BUILD.gn:598]: cc @brave/sec-team. Would you please take a look?

File: patches/components-crx_file-BUILD.gn.patch (hunk 2 above)
Comment (mherrmann) [reply patches/components-crx_file-BUILD.gn.patch:9]: Without this change, the build fails with:

```
ERROR at //build/config/BUILDCONFIG.gn:684:5: Test-only dependency not allowed.
    action("${target_name}") {
    ^-------------------------
//brave/build/mac:crx(//build/toolchain/mac:clang_x64)
which is NOT marked testonly can't depend on
//components/crx_file:crx3_build_action(//build/toolchain/linux:clang_x64)
which is marked testonly. Only targets with "testonly = true"
can depend on other test-only targets.

Either mark it test-only or don't do this dependency.
```

File: build/mac/BUILD.gn (hunk 1 above)
Comment (mherrmann) [reply build/mac/BUILD.gn:598]: I requested a review by opening an issue at https://github.com/brave/reviews/issues/1562.

Comment (diracdeltas) [reply build/mac/BUILD.gn:598]: conclusions:

* this key should be regenerated instead of using the existing crx signing key
* fine to save it as a jenkins credential, load it into this directory for the build, then remove it immediately afterwards

**Review by simonhong — COMMENTED** [review]
Just left trivial comments. Very excited to have O4 in macOS!

File: chromium_src/chrome/browser/ui/webui/help/version_updater_mac.mm [hunk 3]
```diff
@@ -1,209 +1,12 @@
...
-  } else {
-    status_callback_.Run(DISABLED, 0, false, false, std::string(), 0,
-                         std::u16string());
-  }
+#if BUILDFLAG(BRAVE_ENABLE_UPDATER)
+#include "src/chrome/browser/ui/webui/help/version_updater_mac.mm"
 #else
-  status_callback_.Run(DISABLED, 0, false, false, std::string(), 0,
-                       std::u16string());
+#include "chrome/browser/ui/webui/help/version_updater_mac_sparkle.mm"
```
[the remark this answers is not in the document]
Comment (mherrmann) [reply chromium_src/chrome/browser/ui/webui/help/version_updater_mac.mm:11]: It does not! Good catch. I'll make an improvement that fixes this. Thank you!

[the remark this answers is not in the document]
Comment (mherrmann) [reply chromium_src/chrome/browser/ui/webui/help/version_updater_mac.mm]: Fixed in 7af5ba728181dcc7092c69700f6bfef98b684910. What do you think?

[the remark this answers is not in the document]
Comment (mherrmann) [reply chromium_src/chrome/browser/ui/webui/help/version_updater_mac.mm]: (Small whitespace fix in 6c9225ffe9bffc57395833df80a1a1ed3cf436ae.)

File: build/mac/BUILD.gn [hunk 4]
```diff
@@ -412,6 +419,16 @@ action("create_dmg") {
   outputs = [ unsigned_dmg_path ]
 
   deps = [ ":finalize_app" ]
+
+  if (brave_enable_updater) {
+    args += [
+      "--copy",
+      rebase_path(
+          "$root_out_dir/$browser_product_name Packaging/updater/.install",
```
[the remark this answers is not in the document]
Comment (mherrmann) [reply build/mac/BUILD.gn:427]: I don't think so:

```
ERROR at //brave/build/mac/BUILD.gn:424:12: Target not found in this context.
    tmp1 = get_target_outputs("//chrome/updater/mac:browser_install_script")
           ^---------------------------------------------------------------
//chrome/updater/mac:browser_install_script
was not found. get_target_outputs() can only be used for targets
previously defined in the current file.
```

File: chromium_src/chrome/browser/ui/webui/help/version_updater_mac.mm (hunk 3 above)
[the remark this answers is not in the document]
Comment (mherrmann) [reply chromium_src/chrome/browser/ui/webui/help/version_updater_mac.mm]: (And a logic fix in 4e5eec7c58f10b8c4a25959dc1d190622d5b2d31).

File: build/mac/BUILD.gn (hunk 1 above)
Comment (mherrmann) [reply build/mac/BUILD.gn:598]: Regenerated the key in b1362a7c4473b95b136ebf74467e44d07995b47e.

File: patches/components-crx_file-BUILD.gn.patch (hunk 2 above)
Comment (mherrmann) [reply patches/components-crx_file-BUILD.gn.patch:9]: Is this OK for you @goodov?

File: patches/chrome-browser-ui-BUILD.gn.patch [hunk 5]
```diff
@@ -10,7 +10,15 @@ index 768e11b9e56483cbe6235c894bca669b928e4d0c..ba693710323ce67fa923ae9b66346a04
  
    if (enable_vr && is_win) {
      deps += [ "//chrome/browser/vr:vr_base" ]
-@@ -4606,7 +4607,10 @@ static_library("ui") {
+@@ -4477,6 +4478,7 @@ static_library("ui") {
+       "webui/settings/settings_utils_mac.mm",
+       "window_sizer/window_sizer_mac.mm",
+     ]
++    import("//brave/updater/config.gni") if (!brave_enable_updater) { sources -= [ "webui/help/version_updater_mac.mm" ] sources += [ "//brave/chromium_src/chrome/browser/ui/webui/help/version_updater_mac_sparkle.mm" ] }
```
Comment (simonhong) [inline patches/chrome-browser-ui-BUILD.gn.patch:17]: How about this instead of patching and renaming?
```
diff --git a/chromium_src/chrome/browser/ui/webui/help/version_updater_mac.mm b/chromium_src/chrome/browser/ui/webui/help/version_updater_mac.mm
index 88bd1929e85..bb8210e6016 100644
--- a/chromium_src/chrome/browser/ui/webui/help/version_updater_mac.mm
+++ b/chromium_src/chrome/browser/ui/webui/help/version_updater_mac.mm
@@ -3,6 +3,11 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this file,
  * You can obtain one at https://mozilla.org/MPL/2.0/. */

+#include "brave/updater/buildflags.h"
+
+#if BUILDFLAG(BRAVE_ENABLE_UPDATER)
+#include "src/chrome/browser/ui/webui/help/version_updater_mac.mm"
+#else
 #include "brave/chromium_src/chrome/browser/ui/webui/help/version_updater_mac.h"

 #include <memory>
@@ -207,3 +212,4 @@ void VersionUpdaterMac::UpdateStatus(NSDictionary* dictionary) {
 void VersionUpdaterMac::UpdateShowPromoteButton() {
   NOTIMPLEMENTED();
 }
+#endif
```

Comment (mherrmann) [reply patches/chrome-browser-ui-BUILD.gn.patch:17]: This does seem more elegant than what I have. I wonder if it will rebuild when `src/chrome/browser/ui/webui/help/version_updater_mac.mm` is touched. I'll try it! Thanks :)

Comment (mherrmann) [reply patches/chrome-browser-ui-BUILD.gn.patch]: It worked!

Comment (simonhong) [reply patches/chrome-browser-ui-BUILD.gn.patch]: Great! and need deps for this like below as we're including new header :)  

```
diff --git a/chromium_src/chrome/browser/ui/BUILD.gn b/chromium_src/chrome/browser/ui/BUILD.gn
index f33c87c0c66..141cd474d9e 100644
--- a/chromium_src/chrome/browser/ui/BUILD.gn
+++ b/chromium_src/chrome/browser/ui/BUILD.gn
@@ -11,8 +11,13 @@ group("ui") {
   deps = [
     "//brave/components/brave_player/common/buildflags",
     "//brave/components/brave_vpn/common/buildflags",
+    "//brave/updater:buildflags",
   ]

+  if (is_mac) {
+    deps += [ "//brave/updater:buildflags" ]
+  }
+
   if (toolkit_views) {
     deps += [
       "//brave/components/ai_chat/core/common/buildflags",
```

File: patches/components-crx_file-BUILD.gn.patch (hunk 2 above)
Comment (goodov) [reply patches/components-crx_file-BUILD.gn.patch:9]: > Is this OK for you @goodov?

Looks like chrome build this target separately and then runs outside of GN. For our approach we want to run it as a part of GN build flow, so we need to make it non-testonly. Which is fine.

File: patches/chrome-browser-ui-BUILD.gn.patch (hunk 5 above)
Comment (mherrmann) [reply patches/chrome-browser-ui-BUILD.gn.patch]: Fixed in ebda2fff76801f6732942c74c7f33d029a52465a.

**Review by simonhong — APPROVED** [review]
++ for me but please get approve from others also.

File: chromium_src/components/crx_file/crx_verifier.cc [hunk 6]
```diff
@@ -12,29 +12,48 @@
...
+      std::begin(kBraveBrowserUpdatesKeyHash),
+      std::end(kBraveBrowserUpdatesKeyHash));
+  return *brave_browser_updates_key;
+}
+
 // Used in the patch in crx_verifier.cc.
 bool IsBravePublisher(const std::vector<uint8_t>& key_hash) {
-  return GetBravePublisherKeyHash() == key_hash;
+  return key_hash == GetBraveComponentsKeyHash() ||
+         key_hash == GetBraveBrowserUpdatesKeyHash();
```
Comment (goodov) [inline chromium_src/components/crx_file/crx_verifier.cc:48]: wait, this also means that extensions can be signed with any of the keys. I think the idea of creating a separate key for updates was to NOT allow that.

Comment (mherrmann) [reply chromium_src/components/crx_file/crx_verifier.cc:48]: Good point :+1:

Comment (mherrmann) [reply chromium_src/components/crx_file/crx_verifier.cc:48]: I wonder if it will be worth the effort to have separate keys. Per my reading of the code, browser updates, components and extensions all use `crx_file::Verify(...)` to verify signatures. That method has two parameters relevant to signing: The first is the list `required_key_hashes`. It may be empty. But when it isn't, then `Verify(...)` checks that the update binary (a CRX file) is signed with all of the keys in the list. The second relevant parameter is the `VerifierFormat`. It essentially acts as a boolean that determines whether the CRX file also needs to be signed with the "publisher key". The publisher key is hard-coded in the source code and is the key of the browser vendor. In our current implementation, the publisher key is the key that's used to sign component updates. Let's call the publisher key `P`.

Now:

1. Browser updates require `P` (cf. `CrxVerifierFormat()` in `external_constants_default.cc`).
2. Components require `P` (cf. `ToCrxComponent(...)` in `component_updater_service.cc`).
3. (Extensions only require `P` for extensions from the web store. Cf. `verifier_formats.cc`.)

The asked-for requirement is that we have two separate keys for signing browser and component updates.

Given that we already use `P` for components, we could add another key `B` for browser updates. Browser updates would then need to be signed with both `P` and `B`.

One problem is that upstream's code can only create CRX files that are signed with one key. Cf. `crx_creator.cc`. This may sound surprising, given that `crx_file::Verify(...)` takes a list of keys to check. But all uses of the method only ever actually supply 0 or 1 elements to this list.

The second small problem is where we would inject `B` in such a way that it then cannot be used on its own to sign components. A natural place for this seems to be `MakeCrxComponentFromAppInfo`. But this is a pretty general method and it feels wrong to hard-code a signing key inside it. Maybe a solution could be to exploit the fact that the updater is a separate application, so we could actually compile it with a different `P` value. But that too seems somewhat messy, given that `P` is hard-coded in C++.

In short, I feel that the requirement to have separate keys goes against the grain of upstream's implementation. Of course, it should be technically doable. But I wonder if we really want to go through the associated effort.

Comment (mherrmann) [reply chromium_src/components/crx_file/crx_verifier.cc:48]: @goodov I don't know if you will have time, but I would be very interested to hear what you think about my analysis.

Comment (mherrmann) [reply chromium_src/components/crx_file/crx_verifier.cc:48]: I asked upstream; They confirmed 1. - 3.

**Review by iefremov — APPROVED** [review]
lgtm given that sec review is completed and signing related changes are confirmed

File: chromium_src/components/update_client/request_sender.cc [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+constexpr int kBraveKeyVersion = 1;
+constexpr char kBraveKeyPubBytesBase64[] =
+    "MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEMZENJfFz9Jph//JXTejVdn5U+ALz"
+    "NT/Bht/fvkf2hZ5RionWCLzcxmjV3uh0R3MKLfsgI3w7ukou7m8VhkFQSg==";
+
+}  // namespace
+
+namespace client_update_protocol {
+
+class BraveEcdsa : public Ecdsa {
```
[the remark this answers is not in the document]
Comment (mherrmann) [reply chromium_src/components/update_client/request_sender.cc:32]: I added a test in 0d34c02d769cd73fe95c195fa15a36242b862ff0. What do you think?
